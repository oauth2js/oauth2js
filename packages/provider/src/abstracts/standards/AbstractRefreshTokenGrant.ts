import { Mixed } from '@prunus/mixin'

import { InvalidClient } from '../../errors/InvalidClient'
import { InvalidGrant } from '../../errors/InvalidGrant'
import {
  DEFAULT_ACCESS_TOKEN_LIFETIME,
  DEFAULT_REFRESH_TOKEN_LIFETIME,
} from '../../helpers/defaults'
import { Maybe, PromiseToo } from '../../helpers/types'
import { RequestLocation } from '../../interfaces/IRequest'
import { IToken } from '../../interfaces/IToken'
import { Context } from '../../models/Context'
import { UseClientId } from '../../traits/UseClientId'
import { UseTokenByRefreshToken } from '../../traits/UseTokenByRefreshToken'
import { AbstractGrant } from '../AbstractGrant'

export abstract class AbstractRefreshTokenGrant extends Mixed(
  AbstractGrant,
  UseTokenByRefreshToken,
  UseClientId,
) {
  protected accessTokenLifetime: number = DEFAULT_ACCESS_TOKEN_LIFETIME
  protected refreshTokenLifetime: number = DEFAULT_REFRESH_TOKEN_LIFETIME
  protected readonly clientIdLocation: RequestLocation = RequestLocation.Body

  protected readonly refreshTokenLocation: RequestLocation =
    RequestLocation.Body

  protected async createToken(context: Context): Promise<IToken> {
    const token = await this.getOrLoadTokenByRefreshToken(context)

    if ((await this.getClientId(context)) !== token.client.id)
      throw new InvalidClient('Client identifier does not match', context)

    const { client, user, scopes } = token
    const [
      accessToken,
      accessTokenExpiresAt,
      refreshToken,
      refreshTokenExpiresAt,
    ] = await Promise.all([
      this.generateAccessToken(context),
      this.getAccessTokenExpiresAt(context),
      this.generateRefreshToken(context),
      this.getRefreshTokenExpiresAt(context),
    ])

    return {
      client,
      user,
      scopes,
      accessToken,
      accessTokenExpiresAt,
      refreshToken,
      refreshTokenExpiresAt,
    }
  }

  protected async saveToken(token: IToken, context: Context): Promise<IToken> {
    return await this.saveTokenByRefreshedToken(
      token,
      await this._revokeToken(context),
      context,
    )
  }

  private async _revokeToken(context: Context) {
    const tokenToRevoke = await this.getOrLoadTokenByRefreshToken(context)

    if (!(await this.revokeToken(tokenToRevoke, context)))
      throw new InvalidGrant('The token is invalid', context)

    return tokenToRevoke
  }

  protected abstract override getTokenByRefreshToken(
    refreshToken: string,
    context: Context,
  ): PromiseToo<Maybe<IToken>>

  protected abstract saveTokenByRefreshedToken(
    token: IToken,
    refreshedToken: IToken,
    context: Context,
  ): PromiseToo<IToken>

  protected abstract revokeToken(
    token: IToken,
    context: Context,
  ): PromiseToo<boolean>
}
