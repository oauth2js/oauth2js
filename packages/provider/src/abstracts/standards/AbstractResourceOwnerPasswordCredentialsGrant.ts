import { Mixed } from '@prunus/mixin'

import {
  DEFAULT_ACCESS_TOKEN_LIFETIME,
  DEFAULT_REFRESH_TOKEN_LIFETIME,
} from '../../helpers/defaults'
import { Maybe, PromiseToo } from '../../helpers/types'
import { IClient } from '../../interfaces/IClient'
import { RequestLocation } from '../../interfaces/IRequest'
import { IScope } from '../../interfaces/IScope'
import { IToken } from '../../interfaces/IToken'
import { IUser } from '../../interfaces/IUser'
import { Context } from '../../models/Context'
import {
  ClientCredentials,
  UseClientByCredentials,
} from '../../traits/UseClientByCredentials'
import { UseClientById } from '../../traits/UseClientById'
import { UseClientPolymorph } from '../../traits/UseClientPolymorph'
import { UseScopes } from '../../traits/UseScopes'
import {
  UserCredentials,
  UseUserByCredentials,
} from '../../traits/UseUserByCredentials'
import { AbstractGrant } from '../AbstractGrant'

export abstract class AbstractResourceOwnerPasswordCredentialsGrant extends Mixed(
  UseClientPolymorph(UseClientByCredentials, UseClientById),
  AbstractGrant,
  UseUserByCredentials,
  UseScopes,
) {
  protected override readonly type = 'password'
  protected readonly scopeLocation = RequestLocation.Body
  protected readonly refreshTokenLifetime = DEFAULT_REFRESH_TOKEN_LIFETIME
  protected readonly accessTokenLifetime = DEFAULT_ACCESS_TOKEN_LIFETIME
  protected readonly userCredentialsLocation: RequestLocation =
    RequestLocation.Body
  protected readonly clientCredentialsLocation: RequestLocation =
    RequestLocation.Body
  protected readonly clientIdLocation: RequestLocation = RequestLocation.Body

  protected async createToken(context: Context): Promise<IToken> {
    const client = await this.getOrLoadClient(context)
    const scopes = await this.getOrLoadScopesObjects(context)
    const user = await this.loadUser(context)

    const [
      accessToken,
      accessTokenExpiresAt,
      refreshToken,
      refreshTokenExpiresAt,
    ] = await Promise.all([
      this.generateAccessToken(context),
      this.getAccessTokenExpiresAt(context),
      this.generateRefreshToken(context),
      this.getRefreshTokenExpiresAt(context),
    ])

    return {
      accessToken,
      accessTokenExpiresAt,
      refreshToken,
      refreshTokenExpiresAt,
      client,
      scopes,
      user,
    }
  }

  protected abstract override selectUseClientProvider(
    context: Context,
  ): typeof UseClientByCredentials | typeof UseClientById

  protected abstract override getClientByCredentials(
    credentials: ClientCredentials,
    context: Context,
  ): PromiseToo<Maybe<IClient>>

  protected abstract override getClientById(
    clientId: string,
    context: Context,
  ): PromiseToo<Maybe<IClient>>

  protected abstract override getScopesObjects(
    scopes: string[],
    context: Context,
  ): PromiseToo<IScope[]>

  protected abstract override getUserByCredentials(
    credentials: UserCredentials,
    context: Context,
  ): PromiseToo<Maybe<IUser>>

  protected abstract override saveToken(
    token: IToken,
    context: Context,
  ): PromiseToo<IToken>
}
