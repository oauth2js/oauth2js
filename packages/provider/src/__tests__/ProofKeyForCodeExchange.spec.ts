/* eslint-disable max-lines */
import crypto from 'crypto'

import { MockController } from '../../.jest/tools/Controller'
import { Database } from '../../.jest/tools/Database'
import {
  DatabaseAuthorizationCode,
  DatabaseClient,
  DatabaseDeviceCode,
  DatabaseScope,
  DatabaseUser,
} from '../../.jest/tools/DatabaseTypes'
import { AuthorizationCodeGrant } from '../standards/AuthorizationCodeGrant'
import { DeviceCodeGrant } from '../standards/DeviceCodeGrant'
import { ProofKeyForCodeExchange } from '../standards/ProofKeyForCodeExchange'

describe('Proof Key for Code Exchange PKCE', () => {
  it('generate authorization code for a confidential client', async () => {
    const db = Database()

    db.clients.insertOne({
      id: 'client-test-id',
      secret: 'client-test-secret',
      redirectUris: ['http://localhost:8080'],
      type: 'confidential',
    })
    db.users.insertOne({
      username: 'user-test-username',
      password: 'user-test-password',
    })
    db.scopes.insertOne({ alias: 'read', description: 'Read' })
    db.scopes.insertOne({ alias: 'write', description: 'Write' })

    const controller = MockController.from(controller =>
      controller
        .use(ProofKeyForCodeExchange(AuthorizationCodeGrant))
        .implement({
          getClientById: clientId =>
            db.clients.findOne<DatabaseClient>(
              client => client.id === clientId,
            ),
          getAuthorizationCodeByCode: code =>
            db.authorizationCodes.findOne<DatabaseAuthorizationCode>(
              authorizationCode => authorizationCode.code === code,
            ),
          getUserByAuthorizationCode: (
            authorizationCode: DatabaseAuthorizationCode,
          ) => authorizationCode.user,
          authorizeAuthorizationCode: ({ code }, user: DatabaseUser) =>
            void db.authorizationCodes.updateOne<DatabaseAuthorizationCode>(
              { user },
              authorizationCode => authorizationCode.code === code,
            ),
          getScopesObjects: scopes =>
            db.scopes.find<DatabaseScope>(scope =>
              scopes.includes(scope.alias),
            ),
          revokeAuthorizationCode: ({ code }) =>
            !!db.authorizationCodes.softDeleteOne(
              authorizationCode => authorizationCode.code === code,
            ),
          saveAuthorizationCode: authorizationCode =>
            db.authorizationCodes.insertOne(authorizationCode),
          saveTokenByAuthorizationCode: (token, authorizationCode) =>
            db.tokens.insertOne({ ...token, authorizationCode }),
          sendAuthorizationCode(authorizationCode, context) {
            context.response.setHeader('Content-Type', 'text/plain')
            context.response.status(200)
            context.response.send('You autorize this test?')
          },
          isPublicClient: (client: DatabaseClient) => client.type === 'public',
          compareNonPublicClientSecret: (client: DatabaseClient, secret) =>
            client.secret === secret,
        }),
    )

    const challenge = crypto
      .randomBytes(20)
      .toString('hex')
      .split('')
      .map((char, index, { length }) =>
        (index + 1) % 4 === 0 && index + 1 !== length ? `${char}_` : char,
      )
      .join('')

    const response = await controller.release({
      query: {
        response_type: 'code',
        client_id: 'client-test-id',
        client_secret: 'client-test-secret',
        state: 'test-state',
        scope: 'read+write',
        code_challenge: crypto
          .createHash('sha256')
          .update(challenge)
          .digest('base64url'),
      },
    })

    expect(response.status()).toBe(200)
    expect(response.headers).toMatchSnapshot()
    expect(response.body).toBe('You autorize this test?')
    expect(db.authorizationCodes.count()).toBe(1)
    expect(db.authorizationCodes.findOne()).toMatchSnapshot({
      code: expect.any(String),
      codeExpiresAt: expect.any(Date),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      client: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      scopes: [
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      ],
    })
  })

  it('with error when compareNonPublicClientSecret is not implemented', async () => {
    const db = Database()

    db.clients.insertOne({
      id: 'client-test-id',
      secret: 'client-test-secret',
      redirectUris: ['http://localhost:8080'],
      type: 'confidential',
    })
    db.users.insertOne({
      username: 'user-test-username',
      password: 'user-test-password',
    })
    db.scopes.insertOne({ alias: 'read', description: 'Read' })
    db.scopes.insertOne({ alias: 'write', description: 'Write' })

    const controller = MockController.from(controller =>
      controller
        .use(ProofKeyForCodeExchange(AuthorizationCodeGrant))
        .implement({
          getClientById: clientId =>
            db.clients.findOne<DatabaseClient>(
              client => client.id === clientId,
            ),
          getAuthorizationCodeByCode: code =>
            db.authorizationCodes.findOne<DatabaseAuthorizationCode>(
              authorizationCode => authorizationCode.code === code,
            ),
          getUserByAuthorizationCode: (
            authorizationCode: DatabaseAuthorizationCode,
          ) => authorizationCode.user,
          authorizeAuthorizationCode: ({ code }, user: DatabaseUser) =>
            void db.authorizationCodes.updateOne<DatabaseAuthorizationCode>(
              { user },
              authorizationCode => authorizationCode.code === code,
            ),
          getScopesObjects: scopes =>
            db.scopes.find<DatabaseScope>(scope =>
              scopes.includes(scope.alias),
            ),
          revokeAuthorizationCode: ({ code }) =>
            !!db.authorizationCodes.softDeleteOne(
              authorizationCode => authorizationCode.code === code,
            ),
          saveAuthorizationCode: authorizationCode =>
            db.authorizationCodes.insertOne(authorizationCode),
          saveTokenByAuthorizationCode: (token, authorizationCode) =>
            db.tokens.insertOne({ ...token, authorizationCode }),
          sendAuthorizationCode(authorizationCode, context) {
            context.response.setHeader('Content-Type', 'text/plain')
            context.response.status(200)
            context.response.send('You autorize this test?')
          },
          isPublicClient: (client: DatabaseClient) => client.type === 'public',
        }),
    )

    const response = await controller.release({
      query: {
        response_type: 'code',
        client_id: 'client-test-id',
        client_secret: 'client-test-secret',
        state: 'test-state',
        scope: 'read+write',
      },
    })

    expect(response.status()).toBe(302)
    expect(response.headers).toMatchSnapshot()
  })

  it('generate token', async () => {
    const db = Database()

    db.clients.insertOne({
      id: 'client-test-id',
      secret: 'client-test-secret',
      redirectUris: ['http://localhost:8080'],
    })
    db.users.insertOne({
      username: 'user-test-username',
      password: 'user-test-password',
    })
    db.scopes.insertOne({ alias: 'read', description: 'Read' })
    db.scopes.insertOne({ alias: 'write', description: 'Write' })

    const controller = MockController.from(controller =>
      controller
        .use(ProofKeyForCodeExchange(AuthorizationCodeGrant))
        .implement({
          getClientById: clientId =>
            db.clients.findOne<DatabaseClient>(
              client => client.id === clientId,
            ),
          getAuthorizationCodeByCode: code =>
            db.authorizationCodes.findOne<DatabaseAuthorizationCode>(
              authorizationCode => authorizationCode.code === code,
            ),
          getUserByAuthorizationCode: (
            authorizationCode: DatabaseAuthorizationCode,
          ) => authorizationCode.user,
          authorizeAuthorizationCode: ({ code }, user: DatabaseUser) =>
            void db.authorizationCodes.updateOne<DatabaseAuthorizationCode>(
              { user },
              authorizationCode => authorizationCode.code === code,
            ),
          denyAuthorizationCode: ({ code }) =>
            db.authorizationCodes.softDelete(
              authorizationCode => authorizationCode.code === code,
            ),
          getScopesObjects: scopes =>
            db.scopes.find<DatabaseScope>(scope =>
              scopes.includes(scope.alias),
            ),
          revokeAuthorizationCode: ({ code }) =>
            !!db.authorizationCodes.softDeleteOne(
              authorizationCode => authorizationCode.code === code,
            ),
          saveAuthorizationCode: authorizationCode =>
            db.authorizationCodes.insertOne(authorizationCode),
          saveTokenByAuthorizationCode: (token, authorizationCode) =>
            db.tokens.insertOne({ ...token, authorizationCode }),
          sendAuthorizationCode(authorizationCode, context) {
            context.response.setHeader('Content-Type', 'text/plain')
            context.response.status(200)
            context.response.send('You autorize this test?')
          },
          sendErrorWithInvalidClient(error, context) {
            context.response.send('Insecure client')
          },
          sendErrorWithInvalidRedirectUri(error, context) {
            context.response.send('Insecure client redirect')
          },
        }),
    )

    const challenge = crypto
      .randomBytes(20)
      .toString('hex')
      .split('')
      .map((char, index, { length }) =>
        (index + 1) % 4 === 0 && index + 1 !== length ? `${char}_` : char,
      )
      .join('')

    let response = await controller.release({
      query: {
        response_type: 'code',
        client_id: 'client-test-id',
        state: 'test-state',
        scope: 'read+write',
        code_challenge: crypto
          .createHash('sha256')
          .update(challenge)
          .digest('base64url'),
      },
    })

    expect(response.status()).toBe(200)
    expect(response.headers).toMatchSnapshot()
    expect(response.body).toBe('You autorize this test?')
    expect(db.authorizationCodes.count()).toBe(1)
    expect(db.authorizationCodes.findOne()).toMatchSnapshot({
      code: expect.any(String),
      codeExpiresAt: expect.any(Date),
      codeChallenge: expect.any(String),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      client: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      scopes: [
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      ],
    })

    response = await controller.authorize(
      AuthorizationCodeGrant,
      db.authorizationCodes.findOneOrFail(),
      db.users.findOneOrFail(),
    )

    const { code } =
      db.authorizationCodes.findOneOrFail<DatabaseAuthorizationCode>()

    expect(response.status()).toBe(302)
    expect(response.headers).toEqual({
      Location: `http://localhost:8080/?code=${code}&state=test-state`,
    })
    expect(response.body).toBeUndefined()
    expect(db.authorizationCodes.count()).toBe(1)
    expect(db.authorizationCodes.findOne()).toMatchSnapshot({
      code: expect.any(String),
      codeExpiresAt: expect.any(Date),
      codeChallenge: expect.any(String),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      client: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      user: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      scopes: [
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      ],
    })

    response = await controller.token({
      body: {
        grant_type: 'authorization_code',
        code: db.authorizationCodes.findOneOrFail().code,
        redirect_uri: 'http://localhost:8080',
        client_id: 'client-test-id',
        code_verifier: challenge,
      },
    })

    expect(response.status()).toBe(200)
    expect(response.headers).toMatchSnapshot()
    expect(JSON.parse(response.body)).toMatchSnapshot({
      access_token: expect.any(String),
      refresh_token: expect.any(String),
    })
    expect(db.tokens.findOne()).toMatchSnapshot({
      accessToken: expect.any(String),
      accessTokenExpiresAt: expect.any(Date),
      refreshToken: expect.any(String),
      refreshTokenExpiresAt: expect.any(Date),
      scopes: [
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      ],
      client: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      user: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      authorizationCode: {
        code: expect.any(String),
        codeExpiresAt: expect.any(Date),
        codeChallenge: expect.any(String),
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
        deletedAt: expect.any(Date),
        client: {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        user: {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        scopes: [
          {
            createdAt: expect.any(Date),
            updatedAt: expect.any(Date),
          },
          {
            createdAt: expect.any(Date),
            updatedAt: expect.any(Date),
          },
        ],
      },
    })
    expect(db.authorizationCodes.count({ includeDeleted: true })).toBe(1)
    expect(
      db.authorizationCodes.findOne(void 0, { includeDeleted: true }),
    ).toMatchSnapshot({
      code: expect.any(String),
      codeExpiresAt: expect.any(Date),
      codeChallenge: expect.any(String),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      deletedAt: expect.any(Date),
      client: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      user: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      scopes: [
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      ],
    })
  })

  it('pkce for any exchanger', async () => {
    const db = Database()

    db.clients.insertOne({
      id: 'client-test-id',
      secret: 'client-test-secret',
      redirectUris: ['http://localhost:8080'],
    })
    db.users.insertOne({
      username: 'user-test-username',
      password: 'user-test-password',
    })
    db.scopes.insertOne({ alias: 'read', description: 'Read' })
    db.scopes.insertOne({ alias: 'write', description: 'Write' })

    const controller = MockController.from(controller => {
      controller.use(ProofKeyForCodeExchange(DeviceCodeGrant)).implement({
        getClientById: clientId =>
          db.clients.findOne<DatabaseClient>(client => client.id === clientId),
        getScopesObjects: scopes =>
          db.scopes.find<DatabaseScope>(scope => scopes.includes(scope.alias)),
        saveDeviceCode: deviceCode => db.deviceCodes.insertOne(deviceCode),
        pullingInterval: 5 * 1000,
        verificationUri: 'http://localhost:8443',
        getDeviceCodeByCode: code =>
          db.deviceCodes.findOne<DatabaseDeviceCode>(
            deviceCode => deviceCode.code === code,
          ),
        getUserByDeviceCode: (deviceCode: DatabaseDeviceCode) =>
          deviceCode.user ?? false,
        authorizeDeviceCode: ({ code }, user: DatabaseUser) =>
          void db.deviceCodes.updateOne<DatabaseDeviceCode>(
            { user },
            deviceCode => deviceCode.code === code,
          ),
        saveTokenByDeviceCode: (token, deviceCode) =>
          db.tokens.insertOne({ ...token, deviceCode }),
        revokeDeviceCode: ({ code }) =>
          !!db.deviceCodes.softDeleteOne(
            deviceCode => deviceCode.code === code,
          ),
      })
    })

    const challenge = crypto
      .randomBytes(20)
      .toString('hex')
      .split('')
      .map((char, index, { length }) =>
        (index + 1) % 4 === 0 && index + 1 !== length ? `${char}_` : char,
      )
      .join('')

    const { body: rawBody } = await controller.release({
      body: {
        response_type: 'device_code',
        client_id: 'client-test-id',
        client_secret: 'client-test-secret',
        scope: 'read+write',
        code_challenge: crypto
          .createHash('sha256')
          .update(challenge)
          .digest('base64url'),
      },
    })

    const { device_code: code } = JSON.parse(rawBody)

    expect(db.deviceCodes.findOne()).toMatchSnapshot({
      code: expect.any(String),
      codeExpiresAt: expect.any(Date),
      codeChallenge: expect.any(String),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      userCode: expect.any(String),
      client: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      scopes: [
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      ],
    })

    await controller.authorizeOrFail(
      DeviceCodeGrant,
      db.deviceCodes.findOneOrFail<DatabaseDeviceCode>(
        deviceCode => deviceCode.code === code,
      ),
      db.users.findOneOrFail(),
    )

    const response = await controller.token({
      body: {
        grant_type: 'urn:ietf:params:oauth:grant-type:device_code',
        device_code: code,
        client_id: 'client-test-id',
        code_verifier: challenge,
      },
    })

    expect(response.status()).toBe(200)
    expect(response.headers).toMatchSnapshot()
    expect(JSON.parse(response.body)).toMatchSnapshot({
      access_token: expect.any(String),
      refresh_token: expect.any(String),
    })
  })

  it('throws an error when it has no algorithms', () => {
    expect(() => {
      ProofKeyForCodeExchange(AuthorizationCodeGrant, {
        algorithms: {},
      })
    }).toThrowError()
  })
})
