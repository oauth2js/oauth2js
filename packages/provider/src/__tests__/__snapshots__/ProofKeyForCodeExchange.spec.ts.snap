// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Proof Key for Code Exchange PKCE generate authorization code for a confidential client 1`] = `
Object {
  "Content-Type": "text/plain",
}
`;

exports[`Proof Key for Code Exchange PKCE generate authorization code for a confidential client 2`] = `
Object {
  "client": Object {
    "createdAt": Any<Date>,
    "deletedAt": null,
    "id": "client-test-id",
    "redirectUris": Array [
      "http://localhost:8080",
    ],
    "secret": "client-test-secret",
    "type": "confidential",
    "updatedAt": Any<Date>,
  },
  "code": Any<String>,
  "codeExpiresAt": Any<Date>,
  "createdAt": Any<Date>,
  "deletedAt": null,
  "redirectUri": "http://localhost:8080",
  "scopes": Array [
    Object {
      "alias": "read",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Read",
      "updatedAt": Any<Date>,
    },
    Object {
      "alias": "write",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Write",
      "updatedAt": Any<Date>,
    },
  ],
  "state": "test-state",
  "updatedAt": Any<Date>,
}
`;

exports[`Proof Key for Code Exchange PKCE generate token 1`] = `
Object {
  "Content-Type": "text/plain",
}
`;

exports[`Proof Key for Code Exchange PKCE generate token 2`] = `
Object {
  "client": Object {
    "createdAt": Any<Date>,
    "deletedAt": null,
    "id": "client-test-id",
    "redirectUris": Array [
      "http://localhost:8080",
    ],
    "secret": "client-test-secret",
    "updatedAt": Any<Date>,
  },
  "code": Any<String>,
  "codeChallenge": Any<String>,
  "codeChallengeMethod": "S256",
  "codeExpiresAt": Any<Date>,
  "createdAt": Any<Date>,
  "deletedAt": null,
  "redirectUri": "http://localhost:8080",
  "scopes": Array [
    Object {
      "alias": "read",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Read",
      "updatedAt": Any<Date>,
    },
    Object {
      "alias": "write",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Write",
      "updatedAt": Any<Date>,
    },
  ],
  "state": "test-state",
  "updatedAt": Any<Date>,
}
`;

exports[`Proof Key for Code Exchange PKCE generate token 3`] = `
Object {
  "client": Object {
    "createdAt": Any<Date>,
    "deletedAt": null,
    "id": "client-test-id",
    "redirectUris": Array [
      "http://localhost:8080",
    ],
    "secret": "client-test-secret",
    "updatedAt": Any<Date>,
  },
  "code": Any<String>,
  "codeChallenge": Any<String>,
  "codeChallengeMethod": "S256",
  "codeExpiresAt": Any<Date>,
  "createdAt": Any<Date>,
  "deletedAt": null,
  "redirectUri": "http://localhost:8080",
  "scopes": Array [
    Object {
      "alias": "read",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Read",
      "updatedAt": Any<Date>,
    },
    Object {
      "alias": "write",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Write",
      "updatedAt": Any<Date>,
    },
  ],
  "state": "test-state",
  "updatedAt": Any<Date>,
  "user": Object {
    "createdAt": Any<Date>,
    "deletedAt": null,
    "password": "user-test-password",
    "updatedAt": Any<Date>,
    "username": "user-test-username",
  },
}
`;

exports[`Proof Key for Code Exchange PKCE generate token 4`] = `
Object {
  "Content-Type": "application/json",
}
`;

exports[`Proof Key for Code Exchange PKCE generate token 5`] = `
Object {
  "access_token": Any<String>,
  "expires_in": 7200,
  "refresh_token": Any<String>,
  "token_type": "Bearer",
}
`;

exports[`Proof Key for Code Exchange PKCE generate token 6`] = `
Object {
  "accessToken": Any<String>,
  "accessTokenExpiresAt": Any<Date>,
  "authorizationCode": Object {
    "client": Object {
      "createdAt": Any<Date>,
      "deletedAt": null,
      "id": "client-test-id",
      "redirectUris": Array [
        "http://localhost:8080",
      ],
      "secret": "client-test-secret",
      "updatedAt": Any<Date>,
    },
    "code": Any<String>,
    "codeChallenge": Any<String>,
    "codeChallengeMethod": "S256",
    "codeExpiresAt": Any<Date>,
    "createdAt": Any<Date>,
    "deletedAt": Any<Date>,
    "redirectUri": "http://localhost:8080",
    "scopes": Array [
      Object {
        "alias": "read",
        "createdAt": Any<Date>,
        "deletedAt": null,
        "description": "Read",
        "updatedAt": Any<Date>,
      },
      Object {
        "alias": "write",
        "createdAt": Any<Date>,
        "deletedAt": null,
        "description": "Write",
        "updatedAt": Any<Date>,
      },
    ],
    "state": "test-state",
    "updatedAt": Any<Date>,
    "user": Object {
      "createdAt": Any<Date>,
      "deletedAt": null,
      "password": "user-test-password",
      "updatedAt": Any<Date>,
      "username": "user-test-username",
    },
  },
  "client": Object {
    "createdAt": Any<Date>,
    "deletedAt": null,
    "id": "client-test-id",
    "redirectUris": Array [
      "http://localhost:8080",
    ],
    "secret": "client-test-secret",
    "updatedAt": Any<Date>,
  },
  "createdAt": Any<Date>,
  "deletedAt": null,
  "refreshToken": Any<String>,
  "refreshTokenExpiresAt": Any<Date>,
  "scopes": Array [
    Object {
      "alias": "read",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Read",
      "updatedAt": Any<Date>,
    },
    Object {
      "alias": "write",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Write",
      "updatedAt": Any<Date>,
    },
  ],
  "updatedAt": Any<Date>,
  "user": Object {
    "createdAt": Any<Date>,
    "deletedAt": null,
    "password": "user-test-password",
    "updatedAt": Any<Date>,
    "username": "user-test-username",
  },
}
`;

exports[`Proof Key for Code Exchange PKCE generate token 7`] = `
Object {
  "client": Object {
    "createdAt": Any<Date>,
    "deletedAt": null,
    "id": "client-test-id",
    "redirectUris": Array [
      "http://localhost:8080",
    ],
    "secret": "client-test-secret",
    "updatedAt": Any<Date>,
  },
  "code": Any<String>,
  "codeChallenge": Any<String>,
  "codeChallengeMethod": "S256",
  "codeExpiresAt": Any<Date>,
  "createdAt": Any<Date>,
  "deletedAt": Any<Date>,
  "redirectUri": "http://localhost:8080",
  "scopes": Array [
    Object {
      "alias": "read",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Read",
      "updatedAt": Any<Date>,
    },
    Object {
      "alias": "write",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Write",
      "updatedAt": Any<Date>,
    },
  ],
  "state": "test-state",
  "updatedAt": Any<Date>,
  "user": Object {
    "createdAt": Any<Date>,
    "deletedAt": null,
    "password": "user-test-password",
    "updatedAt": Any<Date>,
    "username": "user-test-username",
  },
}
`;

exports[`Proof Key for Code Exchange PKCE pkce for any exchanger 1`] = `
Object {
  "client": Object {
    "createdAt": Any<Date>,
    "deletedAt": null,
    "id": "client-test-id",
    "redirectUris": Array [
      "http://localhost:8080",
    ],
    "secret": "client-test-secret",
    "updatedAt": Any<Date>,
  },
  "code": Any<String>,
  "codeChallenge": Any<String>,
  "codeChallengeMethod": "S256",
  "codeExpiresAt": Any<Date>,
  "createdAt": Any<Date>,
  "deletedAt": null,
  "scopes": Array [
    Object {
      "alias": "read",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Read",
      "updatedAt": Any<Date>,
    },
    Object {
      "alias": "write",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Write",
      "updatedAt": Any<Date>,
    },
  ],
  "updatedAt": Any<Date>,
  "userCode": Any<String>,
}
`;

exports[`Proof Key for Code Exchange PKCE pkce for any exchanger 2`] = `
Object {
  "Content-Type": "application/json",
}
`;

exports[`Proof Key for Code Exchange PKCE pkce for any exchanger 3`] = `
Object {
  "access_token": Any<String>,
  "expires_in": 7200,
  "refresh_token": Any<String>,
  "token_type": "Bearer",
}
`;

exports[`Proof Key for Code Exchange PKCE with error when compareNonPublicClientSecret is not implemented 1`] = `
Object {
  "Location": "http://localhost:8080/?error=server_error&error_description=The+%60compareNonPublicClientSecret%60+definition+is+missing+from+the+implementation.",
}
`;
